File: Leetcode 3701-3800/3786/3786.py
# Leetcode 3786: Total Sum of Interaction Cost in Tree Groups
# https://leetcode.com/problems/total-sum-of-interaction-cost-in-tree-groups/
# Solved on 24th of December, 2025
class Solution:
    def interactionCost(self, n: int, edges: list[list[int]], group: list[int]) -> int:
        """
        Calculates the total sum of interaction costs in tree groups.

        Args:
            n: The number of nodes in the tree.
            edges: A list of lists representing the edges of the tree.
            group: A list where group[i] is the group ID of node i.

        Returns:
            The total sum of interaction costs.
        """

        if n == 1:
            return 0

        adjacencyList = [[] for _ in range(n)]
        for u, v in edges:
            adjacencyList[u].append(v)
            adjacencyList[v].append(u)

        groupCounts = [0] * 21
        for g in group:
            groupCounts[g] += 1

        parent = [-1] * n
        order = []
        stack = [0]

        while stack:
            u = stack.pop()
            order.append(u)
            for v in adjacencyList[u]:
                if v != parent[u]:
                    parent[v] = u
                    stack.append(v)

        subtreeCounts = [[0] * 21 for _ in range(n)]
        totalInteractionCost = 0

        for i in range(n - 1, -1, -1):
            u = order[i]
            currentGroup = group[u]
            subtreeCounts[u][currentGroup] += 1

            for v in adjacencyList[u]:
                if v == parent[u]:
                    continue

                for groupIndex in range(1, 21):
                    countInSubtree = subtreeCounts[v][groupIndex]
                    if countInSubtree > 0:
                        countOutside = groupCounts[groupIndex] - countInSubtree
                        totalInteractionCost += countInSubtree * countOutside
                        subtreeCounts[u][groupIndex] += countInSubtree

        return totalInteractionCost

File: Leetcode 3701-3800/3786/3786.java
// Leetcode 3786: Total Sum of Interaction Cost in Tree Groups
// https://leetcode.com/problems/total-sum-of-interaction-cost-in-tree-groups/
// Solved on 24th of December, 2025
import java.util.ArrayList;
import java.util.List;

class Solution {
    private long totalInteractionCost;
    private List<List<Integer>> adjacencyList;
    private int[][] subtreeGroupCounts;
    private int[] totalGroupCounts;
    private int[] nodeGroups;

    /**
     * Calculates the total sum of interaction costs in tree groups.
     * @param n The number of nodes in the tree.
     * @param edges An array of edges representing the tree structure.
     * @param group An array where group[i] is the group ID of node i.
     * @return The total sum of interaction costs.
     */
    public long interactionCosts(int n, int[][] edges, int[] group) {
        totalInteractionCost = 0;
        nodeGroups = group;
        
        adjacencyList = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adjacencyList.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            adjacencyList.get(edge[0]).add(edge[1]);
            adjacencyList.get(edge[1]).add(edge[0]);
        }

        totalGroupCounts = new int[22];
        for (int g : group) {
            totalGroupCounts[g]++;
        }

        subtreeGroupCounts = new int[n][22];
        dfs(0, -1);

        return totalInteractionCost;
    }

    private void dfs(int currentNode, int parentNode) {
        int currentGroup = nodeGroups[currentNode];
        subtreeGroupCounts[currentNode][currentGroup] = 1;

        for (int neighbor : adjacencyList.get(currentNode)) {
            if (neighbor == parentNode) {
                continue;
            }

            dfs(neighbor, currentNode);

            for (int i = 1; i <= 20; i++) {
                subtreeGroupCounts[currentNode][i] += subtreeGroupCounts[neighbor][i];
            }
        }

        if (parentNode != -1) {
            for (int i = 1; i <= 20; i++) {
                long countInSubtree = subtreeGroupCounts[currentNode][i];
                long countOutsideSubtree = totalGroupCounts[i] - countInSubtree;
                totalInteractionCost += countInSubtree * countOutsideSubtree;
            }
        }
    }
}